celery/canvas.py

```diff
@@ -663,8 +662,16 @@ def run(self, args=None, kwargs=None, group_id=None, chord=None,
             first_task = tasks.pop()
             options = _prepare_chain_from_options(options, tasks, use_link)
 
-            first_task.apply_async(**options)
-            return results[0]
+            real_result = first_task.apply_async(**options)
+            # If we only have a single task, it may be important that we pass
+            # the real result object rather than the one obtained via freezing.
+            # e.g. For `GroupResult`s, we need to pass back the result object
+            # which will actually have its promise fulfilled by the subtasks,
+            # something that will never occur for the frozen result.
+            if not tasks:
+                return real_result
```

Comment:
line #L672 is not covered by tests yet

Reply:
This has been pulled out into #6411 - it'll disappear from the diff when I next push